File: backend/models/question.go
package models

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type QuestionDifficulty int

const (
	EasyDifficulty QuestionDifficulty = iota + 1
	MediumDifficulty
	HardDifficulty
)

type ConfidenceLevel int

const (
	VeryLowConfidence ConfidenceLevel = iota + 1
	LowConfidence
	MediumConfidence
	HighConfidence
	VeryHighConfidence
)

type Question struct {
	ID          int                `json:"id"`
	Tags        []string           `json:"tags"`
	Title       string             `json:"title"`
	Slug        string             `json:"slug"`
	Description string             `json:"description"`
	Difficulty  QuestionDifficulty `json:"difficulty"`
}

type QuestionTag struct {
	ID         int    `json:"id"`
	QuestionID int    `json:"questionId"`
	Tag        string `json:"tag"`
}

type QuestionSubmission struct {
	ID              uuid.UUID       `json:"id"`
	QuestionID      int             `json:"questionId"`
	Date            time.Time       `json:"date"`
	ConfidenceLevel ConfidenceLevel `json:"confidenceLevel"`
}

type QuestionSubmissionUserStats struct {
	ID               uuid.UUID     `json:"id"`
	QuestionID       int           `json:"questionID"`
	UserID           uuid.UUID     `json:"userID"`
	NumOfSubmissions uint          `json:"numOfSubmissions"`
	AvgDuration      time.Duration `json:"avgDuration"`
}

func DetermineDifficulty(val int) (QuestionDifficulty, error) {
	if val < int(EasyDifficulty) && val > int(HardDifficulty) {
		return EasyDifficulty, fmt.Errorf("%d is not recognized as a valid difficulty level", val)
	}

	return QuestionDifficulty(val), nil
}

File: frontend/src/models/Question.ts
export const ConfidenceLevel = {
    Again: 1,
    Hard: 2,
    Good: 3,
    Easy: 4,
} as const;
export type ConfidenceLevel = typeof ConfidenceLevel[keyof typeof ConfidenceLevel];

const confidenceLevelStrings: Record<ConfidenceLevel, string> = {
    [ConfidenceLevel.Again]: 'again',
    [ConfidenceLevel.Hard]: 'hard',
    [ConfidenceLevel.Good]: 'good',
    [ConfidenceLevel.Easy]: 'easy',
}
export function confidenceLevelToString(level: ConfidenceLevel): string {
    return confidenceLevelStrings[level]
}

export type QuestionSubmissionWithDetails = {
    id: number;
    questionId: number;
    confidenceLevel: string;
    timeTaken: number | null;
    question: {
        id: number;
        title: string;
        slug: string;
        difficulty: string;
        tags: string[];
    };
}
